prov/gni/src/gnix_cq.c

```diff
@@ -528,48 +521,53 @@ DIRECT_FN STATIC ssize_t gnix_cq_readfrom(struct fid_cq *cq, void *buf,
 	return read_count ?: -FI_EAGAIN;
 }
 
-DIRECT_FN STATIC ssize_t gnix_cq_read(struct fid_cq *cq, void *buf, size_t count)
+static ssize_t __gnix_cq_sreadfrom(int blocking, struct fid_cq *cq, void *buf,
+				     size_t count, fi_addr_t *src_addr,
+				     const void *cond, int timeout)
 {
-	return gnix_cq_readfrom(cq, buf, count, NULL);
-}
+	struct gnix_fid_cq *cq_priv;
 
-DIRECT_FN STATIC ssize_t gnix_cq_sreadfrom(struct fid_cq *cq, void *buf,
-					   size_t count, fi_addr_t *src_addr,
-					   const void *cond, int timeout)
-{
-	double start_ms = 0.0, end_ms = 0.0;
-	int ret;
+	cq_priv = container_of(cq, struct gnix_fid_cq, cq_fid);
+	if ((blocking && !cq_priv->wait) ||
+	    (blocking && cq_priv->attr.wait_obj == FI_WAIT_SET))
+		return -FI_EINVAL;
 
-	ret = gnix_cq_readfrom(cq, buf, count, NULL);
-	if (ret != -FI_EAGAIN || timeout == 0)
-		return ret;
+	if (_gnix_queue_peek(cq_priv->errors))
+		return -FI_EAVAIL;
 
-	if (timeout > 0)
-		start_ms = fi_gettime_ms();
+	if (cq_priv->wait)
```

Comment:
We could be inside a gnix_cq_read() (non-blocking) even if the CQ has a wait set bound to it. This check would do the wrong thing in that case.